### ISSUE TITLE
UndirectedSpecifics.getEdgeContainer should use computeIfAbsent

### ISSUE THREAD

**Issue by shevek** [jgrapht/jgrapht#958]
```
 * JGraphT version: github master
 * Java version (java -version)/platform:  
```

**Issue**
https://github.com/jgrapht/jgrapht/blob/master/jgrapht-core/src/main/java/org/jgrapht/graph/specifics/UndirectedSpecifics.java#L287

If we add a Function/Supplier field to UndirectedSpecifics, then getEdgeContainer can use Map.computeIfAbsent, which is much faster.

My app is spending 99% of its time in HashMap.get() when trying to run GnpRandomGraphGenerator(10_000, 0.2)

FWIW I'm not that impressed with UndirectedEdgeContainer:
* I can't do the computeIfAbsent in my subclass of Specifics because the constructor of UndirectedEdgeContainer is package protected.
* I'll gladly pay the cost of losing the UnmodifiableEdgeSet if, in return, I also lose the overhead of allocating that object.

**Steps to reproduce (small coding example)**

        UndirectedGraph graph = ... (our private Fast implementation)
        GnpRandomGraphGenerator<LineageNode, EntityRelationshipEdge> generator = new GnpRandomGraphGenerator<>(10_000, 0.2);
        generator.generateGraph(graph, null);

run visualvm/some profiler.

**Expected behaviour**

Faster.

**Other information**

I'm sorry I raise such awful, incomplete tickets.

**Comment by shevek** [comment]
Of course I can (and will, when fewer people are screaming at me) implement a complete performance-oriented Specifics class, since there's nothing in the interface I can't do myself.

**Comment by jsichi** [comment]
Couldn't we just use computeIfAbsent as is?  Separately, I can understand the desire to have a Supplier/Function (or to just extend the EdgeSetFactory interface with another method which produces the container).

I agree that we ought to make the XxxEdgeContainer constructors protected to match the fact that the classes are public.

The option to take off your seatbelt and skip the overhead of the unmodifiable sets seems good too.

@d-michail @jkinable any opinions on these?

**Comment by jsichi** [comment]
(For computeIfAbsent, perhaps you mean that a private field is necessary in order to avoid the overhead of reconstructing the Function object every time?  If that's the case, then yes, that makes sense.  On first read I thought you were suggesting a parameter.)

**Comment by shevek** [comment]
Yes, just a private field, and use computeIfAbsent as is. I'm just trying to make the call be allocation free as well, hence the field. A field per graph seems cheap, and visiting hash tables is surprisingly expensive when everything else is tuned up. I haven't quite wrapped my head around the memory allocation behaviour of static lambdas yet, but this one isn't static anyway.

**Comment by HannesWell** [comment]
I also run the case you mentioned initially and sampled and profiled it and was surprised that   `FastLookupUndirectedSpecifics.createEdgeToTouchingVerticesIfAbsent()` consumes around 80% (says the sampler) of the entire runtime of the case.

Just from looking at the code I expected using a `HashMap` instead of an `ArrayList` to get the vertex at a certain index, would have a major performance impact (which is why I created #1023 ). But it turned out that calling `HashMap.get()` directly in the `generateGraph()` method consumed only around 2% of the runtime  (according to the sampler).

Also from a quick comparison I found that my change proposed in #1021 almost did not affect the runtime of the graph-generation, which I also didn't expect.
I will try it again with a more reliable and meaningful benchmark (1. using a more predictable generation schema like the CompleteGraphGenerator and 2. doing warm-up and repetitions) in the next days.

But what I can say for sure is that using Map.computeIfAbsent() in `UndirectedSpecifics.getEdgeContainer(V)` will not have any effect, because the case were `ec` is null does not occur in the test-case. I'm not sure, but as far as I can oversee it, `ec` is never null, because this would only be the case if the vertex were not added to the graph before (which is usually checked in the calling methods, which is why I added a corresponding comment in #1021). But correct me if I'm wrong.

Regarding the question if the Function (lambda) object should be saved in a private field:
For non capturing lambdas the answer is definitely no, because the JVM creates non caputring-lambdas as singleton and reuses them. For lambdas capturing local variables the answer is also no, simply because it isn't possible.
The case in question here, were the reference of a field of the containing class is captured, is not so clear. But from this stackoverflow-question (https://stackoverflow.com/questions/23983832/is-method-reference-caching-a-good-idea-in-java-8/23991339#23991339) I also tend to _no_.
This blog concluded that the performance is almost not affected at all: https://medium.com/@damian.kolasa/performance-implications-of-lambdas-and-method-references-when-mapping-a-stream-in-java-79f6e2da6806
Additionally the JVM has great freedoms how it handles lambdas, so it could be the case that lambda-creation is further improved in the future.

Nevertheless if the case where caching a lambda could be useful remains, I will try and benchmark it.

Coming back to the actual problem: I think the performance of adding edges could be further improved in `FastLookupUndirectedSpecifics`, by integrating/inlining the single methods tighter into each other. So Map's compute methods can be used (which should reduce the number of map lookups) and creation of multiple UnorderedPair-objects can be avoided.
I will try this and benchmark it and then will report the results.
